### PR TITLE
Enable file locking in musl stdio

### PR DESCRIPTION
**Pull request by sbc100**
This is second attempt at landing a version of
https://github.com/emscripten-core/emscripten/commit/d5d5f69ad066aee2e22a601921411b94206d035d

The first time we tried it #13837 we ran into issues with
test_pthread_exit_process deadlocking which I tracked down to an issue
with `system/lib/libc/musl/src/thread/__wait.c` where it was blocking
the main thread forever rather then looping and calling
`emscripten_main_thread_process_queued_calls`.

Includes a version of #14481 so that should land before this does.

Fixes: #13194

### CODE REVIEW

**Comment by kleisauke** [comment]
Looks like the test failure in `wasm3.test_pthread_abort` (https://circleci.com/gh/emscripten-core/emscripten/455648) is the same as the one reported in #15014.

File: system/lib/pthread/pthread_create.c [hunk 1]
```diff
@@ -61,6 +71,17 @@ int __pthread_create(pthread_t *restrict res, const pthread_attr_t *restrict att
     return EINVAL;
   }
 
+  pthread_t self = __pthread_self();
```
Comment (kleisauke) [inline system/lib/pthread/pthread_create.c:74]: This variable is unused now. I think it was probably used in musl to set the thread specific data to `__pthread_tsd_main` when the first thread is created.
https://github.com/emscripten-core/emscripten/blob/4d864df0a57024d667e8db171aab2c3385d04c30/system/lib/libc/musl/src/thread/pthread_create.c#L201

**Review by kleisauke — APPROVED** [review]
Thanks for keeping track of this! This looks good to me.

**Comment by kripken** [comment]
> Includes a version of #14481 so that should land before this does.

If you also need my review, perhaps this PR could target that other PR? Then I'd just see the diff between them. (As I'm not familiar with the other PR, it would be hard for me to just review the new parts here, and I don't see an easy way on github to get that diff.) But @kleisauke 's review is probably enough already?

**Pull request merged**